go/vt/discovery/topology_watcher_test.go

```diff
@@ -506,3 +506,141 @@ func TestFilterByKeyspace(t *testing.T) {
+
+	// Add a second tablet to the topology that should get filtered out by the keyspace filter
+	tablet2 := &topodatapb.Tablet{
+		Alias: &topodatapb.TabletAlias{
+			Cell: "aa",
+			Uid:  2,
+		},
+		Hostname: "host2",
+		PortMap: map[string]int32{
+			"vt": 789,
+		},
+		Keyspace: "ks3",
+		Shard:    "shard",
+	}
+	if err := ts.CreateTablet(context.Background(), tablet2); err != nil {
+		t.Fatalf("CreateTablet failed: %v", err)
+	}
+	tw.loadTablets()
+
+	counts = checkOpCounts(t, counts, map[string]int64{"ListTablets": 1, "GetTablet": 1})
+	checkChecksum(t, tw, 2762153755)
+
+	// Check the new tablet is NOT returned by GetAllTablets().
+	allTablets = fhc.GetAllTablets()
+	key = TabletToMapKey(tablet2)
+	if _, ok := allTablets[key]; ok || len(allTablets) != 1 {
+		t.Errorf("fhc.GetAllTablets() = %+v; want %+v", allTablets, tablet2)
+	}
+
+	// Load the tablets again to show that when refreshKnownTablets is disabled,
+	// only the list is read from the topo and the checksum doesn't change
+	tw.loadTablets()
+	counts = checkOpCounts(t, counts, map[string]int64{"ListTablets": 1})
+	checkChecksum(t, tw, 2762153755)
+
+	// same tablet, different port, should update (previous
+	// one should go away, new one be added)
+	origTablet := proto.Clone(tablet).(*topodatapb.Tablet)
+	origKey := TabletToMapKey(tablet)
+	tablet.PortMap["vt"] = 456
+	if _, err := ts.UpdateTabletFields(context.Background(), tablet.Alias, func(t *topodatapb.Tablet) error {
+		t.PortMap["vt"] = 456
+		return nil
+	}); err != nil {
+		t.Fatalf("UpdateTabletFields failed: %v", err)
+	}
+	tw.loadTablets()
+	allTablets = fhc.GetAllTablets()
+	key = TabletToMapKey(tablet)
+
+	counts = checkOpCounts(t, counts, map[string]int64{"ListTablets": 1})
+	if _, ok := allTablets[origKey]; !ok || len(allTablets) != 1 || !proto.Equal(allTablets[origKey], origTablet) {
+		t.Errorf("fhc.GetAllTablets() = %+v; want %+v", allTablets, origTablet)
+	}
+	if _, ok := allTablets[key]; ok {
+		t.Errorf("fhc.GetAllTablets() = %+v; don't want %v", allTablets, key)
+	}
+	checkChecksum(t, tw, 2762153755)
+
+	// Remove the tracked tablet from the topo and check that it is detected as being gone.
+	if err := ts.DeleteTablet(context.Background(), tablet.Alias); err != nil {
+		t.Fatalf("DeleteTablet failed: %v", err)
+	}
+	if _, err := topo.FixShardReplication(context.Background(), ts, logger, "aa", "ks1", "shard"); err != nil {
+		t.Fatalf("FixShardReplication failed: %v", err)
+	}
+	tw.loadTablets()
+	counts = checkOpCounts(t, counts, map[string]int64{"ListTablets": 1, "RemoveTablet": 1})
+	checkChecksum(t, tw, 789108290)
+
+	// confirm state of topo
+	allTablets = fhc.GetAllTablets()
+	key = TabletToMapKey(tablet)
+	if _, ok := allTablets[key]; ok || len(allTablets) != 0 {
+		t.Errorf("fhc.GetAllTablets() = %+v; don't want %v", allTablets, key)
+	}
+	key = TabletToMapKey(tablet2)
+	if _, ok := allTablets[key]; ok || len(allTablets) != 0 {
+		t.Errorf("fhc.GetAllTablets() = %+v; want %+v", allTablets, tablet2)
+	}
```

Comment:
Similar comments here that we can be more precise about what we're really testing for in the conditionals (only including the key ones) and in the failing log message.